### PR TITLE
directory-menu@torchipeppo: Change to TextIconApplet and other improvements

### PR DESCRIPTION
**Pull request by anaximeno**
Hi @torchipeppo 

I made some changes here:

1. Turned the applet from an IconApplet to a TextIconApplet this way it will be possible to display a text label in the front of the icon as shown in the image below:

Captura de ecrã de 2024-04-07 17-38-14

Captura de ecrã de 2024-04-07 17-38-54

2. Updated the Character Limit setting to show a switch instead to enable the character limiting, and only when it is enabled that the user has to define the character limit (the logic on the Python script continues the same)

Captura de ecrã de 2024-04-07 17-44-29

3. I also removed the tooltip text entry in favor of the panel text added in 1. and instead made it static for "Directory Menu" which is the title of the applet

Please tell me if you approve these changes.

### CODE REVIEW

**Comment by rcalixte** [comment]
> 3\. I also removed the tooltip text entry in favor of the panel text added in 1. and instead made it static for "Directory Menu" which is the title of the applet

I actually liked this because it helped differentiate multiple instances on the same panel. It would be nice to have some flexibility here.

**Comment by rcalixte** [comment]
> 1. Turned the applet from an IconApplet to a TextIconApplet this way it will be possible to display a text label in the front of the icon as shown in the image below:

Can you add a screenshot of the before and after? I guess I'm not seeing the difference quite yet.

**Comment by anaximeno** [comment]
> > 3. I also removed the tooltip text entry in favor of the panel text added in 1. and instead made it static for "Directory Menu" which is the title of the applet
> 
> I actually liked this because it helped differentiate multiple instances on the same panel. It would be nice to have some flexibility here.

Wouldn't the change in 1. help with that?

**Comment by anaximeno** [comment]
@rcalixte I improved the screenshots with a before and after comparison for better comprehension of the changes

**Comment by rcalixte** [comment]
> > > 3. I also removed the tooltip text entry in favor of the panel text added in 1. and instead made it static for "Directory Menu" which is the title of the applet
> > 
> > 
> > I actually liked this because it helped differentiate multiple instances on the same panel. It would be nice to have some flexibility here.
> 
> Wouldn't the change in 1. help with that (For example if each instance starts from one place you could change the panel text with the name of the start place)?

Maybe make them independent options?

**Comment by anaximeno** [comment]
> > > > 3. I also removed the tooltip text entry in favor of the panel text added in 1. and instead made it static for "Directory Menu" which is the title of the applet
> > > 
> > > 
> > > I actually liked this because it helped differentiate multiple instances on the same panel. It would be nice to have some flexibility here.
> > 
> > 
> > Wouldn't the change in 1. help with that (For example if each instance starts from one place you could change the panel text with the name of the start place)?
> 
> Maybe make them independent options?

They aren't really "dependent", it's just that it didn't seem necessary after the change in 1. But since it seems that, there's a case where you would find it useful (even with the option in 1), then I will add that back, is that ok?

**Comment by rcalixte** [comment]
> They aren't really "dependent", it's just that it didn't seem necessary after the change in 1. But since it seems that, there's a case where you would find it useful (even with the option in 1), then I will add that back, is that ok?

I may just be confused.

**Comment by anaximeno** [comment]
> I may just be confused.

As a quick summary, originally when you added the directory-menu applet to the panel, it would only show an icon corresponding to the applet, what I have done was change the type of the applet to a TextIcon applet which is an applet that can display both an icon and (optionally) some text with the icon in the panel.

**Comment by rcalixte** [comment]
> As a quick summary, originally when you added the directory-menu applet to the panel, it would only show an icon corresponding to the applet, what I have done was change the type of the applet to a TextIcon applet which is an applet that can display both an icon and (optionally) some text with the icon in the panel.

Ah yeah, for some reason I thought it already had panel text as an option.

Unrelated, how do you (and @torchipeppo) feel about an option for a key binding that opens the menu? It would default to empty of course but it would help for folks who want to navigate solely via the keyboard.

**Comment by anaximeno** [comment]
> Unrelated, how do you (and @torchipeppo) feel about an option for a key binding that opens the menu? It would default to empty of course but it would help for folks who want to navigate solely via the keyboard.

I'm in favor of that, since you can also navigate through menus and submenus with the keyboard, it would make sense to have a key binding open and close it too.

**Comment by rcalixte** [comment]
> I'm in favor of that, since you can also navigate through menus and submenus with the keyboard, it would make sense to have a key binding to open and close it too.

The Escape key already closes the menu so we would just need a binding for the open menu function I think.

**Comment by torchipeppo** [comment]
@anaximeno @rcalixte 
Let me go through the points one by one:

> Turned the applet from an IconApplet to a TextIconApplet

I can see how that would help in certain setups, so I think it's a good addition. I will require that the label is optional *(I assume you just set it to the empty string?)* and the configurable tooltip remains, though, if only because I expect an extra text label to be detrimental for my personal use case. But, again, this is part of a community, and I always like seeing a range of customization options on the stuff I use, so the more people this can serve, the better.

So, in conclusion, more customization is good as long as it doesn't break the current look.

> Character limit switch

Nice! The "-1 to disable" thing has always been nothing more than a quick way to get the job done.

> Improve the adaptation to vertical panels

I don't think this point is further explained? I'll have to check the code.

> An option for a key binding that opens the menu

I'm also in favor.

File: directory-menu@torchipeppo/files/directory-menu@torchipeppo/applet.js [hunk 1]
```diff
@@ -36,15 +49,21 @@ class CassettoneApplet extends Applet.IconApplet {
...
-        this.set_applet_icon_symbolic_name(this.icon_name);
         this.set_applet_tooltip(_(this.tooltip_text));
+        this.set_applet_icon_symbolic_name(this.icon_name);
+        this.set_show_label_in_vertical_panels(false);
+        this.set_applet_label(this.label)
 
         this.actor.connect('enter-event', Lang.bind(this, this.on_enter_event));
         this.actor.connect('button-release-event', Lang.bind(this, this.on_button_release_event));
+
+		this.setAllowedLayout(Applet.AllowedLayout.BOTH);
```
Comment (torchipeppo) [inline directory-menu@torchipeppo/files/directory-menu@torchipeppo/applet.js:66]: I still don't understand the "improved vertical panels" thing. Is it this line?

**Review by torchipeppo — APPROVED** [review]

File: directory-menu@torchipeppo/files/directory-menu@torchipeppo/applet.js (hunk 1 above)
Comment (anaximeno) [reply directory-menu@torchipeppo/files/directory-menu@torchipeppo/applet.js:66]: Yes, generally, when you moved the applet to a vertical panel it would alert to the user that it might not work well there, but for what I have tested it works without any issues in a vertical panel, also theres one more line above that disables the text on vertical panels if set.

**Comment by anaximeno** [comment]
@torchipeppo thank you for your points. The text label is set to an empty string by default (in the image above I decided to put a text there to show how it would work), and after taking with @rcalixte we decided that the tooltip remains (there are points on why I decided to remove an add it again in the comments above) and improving the adaption to the vertical panel is indeed related to the line you mentioned in the code review below. In conclusion the default look (the one you get after the installation) keeps the same, only the user will have the opportunity to add a text label to the panel if they want to, and the character limiting will work a bit differently.

**Comment by torchipeppo** [comment]
@anaximeno It's all good from my side then.

Thank you for your contribution! ^_^

**Review by rcalixte — APPROVED** [review]
